### ISSUE TITLE
[Streaming] Replace Mock Tensor Conversion Functions with Real Neural Network Operations

### ISSUE THREAD

**Issue by EffortlessSteven** [EffortlessMetrics/BitNet-rs#310]
# [Streaming] Replace Mock Tensor Conversion Functions with Real Neural Network Operations

## Problem Description

The streaming generation system in `bitnet-inference` contains critical mock implementations that prevent real neural network inference from working correctly. Specifically, the `tokens_to_tensor` and `tensor_to_logits` functions in the `forward_pass` method are stub implementations that create mock tensors instead of performing actual tensor operations required for BitNet neural network inference.

## Environment

- **Location**: `/home/steven/code/Rust/BitNet-rs/crates/bitnet-inference/src/streaming.rs`
- **Functions**: `tokens_to_tensor` (lines 444-447), `tensor_to_logits` (lines 449-453)
- **Context**: `GenerationStream::forward_pass` method called during streaming token generation
- **Rust Version**: 1.90.0 (Rust 2024 edition)
- **Affected Features**: All streaming inference (`--features cpu`, `--features gpu`)

### Mock Tensor Creation
```rust
// Line 444-447: Mock implementation
fn tokens_to_tensor(tokens: &[u32]) -> Result<ConcreteTensor> {
    Ok(ConcreteTensor::mock(vec![1, tokens.len()]))
}
```

**Problems:**
- Creates mock tensors with arbitrary dimensions instead of proper token embeddings
- No actual tokenization or embedding lookup
- Dimensions don't match neural network input requirements
- Breaks the entire inference pipeline with fake data

### Mock Logits Extraction
```rust
// Line 449-453: Mock implementation
fn tensor_to_logits(_tensor: &ConcreteTensor, vocab_size: usize) -> Result<Vec<f32>> {
    // In a real implementation, this would extract logits from the tensor
    Ok(vec![0.1; vocab_size])
}
```

**Problems:**
- Returns uniform logits (0.1 for all tokens) instead of extracting from neural network output
- Ignores the actual tensor computation results
- Prevents meaningful token sampling and generation
- Results in deterministic, low-quality output regardless of input

## Root Cause Analysis

The mock implementations were likely created as placeholders during initial development but were never replaced with real tensor operations. This creates a fundamental disconnect between the streaming infrastructure (which is production-ready) and the actual neural network computation (which is mocked).

**Impact Assessment:**
- **Severity**: Critical - prevents any meaningful text generation
- **Affected Users**: All users attempting streaming inference
- **Performance Impact**: High - no real computation occurs, but quality is zero
- **Business Impact**: Blocks production deployment of streaming inference

## Reproduction Steps

1. Build with streaming features: `cargo build --no-default-features --features cpu`
2. Attempt streaming inference with any model:
   ```bash
   cargo run -p xtask -- infer --model path/to/model.gguf --prompt "Hello world"
   ```
3. **Expected**: Meaningful text generation based on model and prompt
4. **Actual**: Random/uniform text generation due to mock logits (all 0.1 values)

**Evidence**: All generated tokens will have identical sampling probabilities due to uniform logits.

## Proposed Solution

Replace the mock implementations with real tensor operations that integrate with the existing `ConcreteTensor` enum and `BitNetTensor` infrastructure.

#### Real Token-to-Tensor Conversion
```rust
/// Convert tokens to input tensor for neural network inference
fn tokens_to_tensor(tokens: &[u32]) -> Result<ConcreteTensor> {
    // Create tensor from token IDs with proper shape for model input
    let shape = vec![1, tokens.len()]; // [batch_size, sequence_length]

    // Convert to proper tensor format
    let data: Vec<f32> = tokens.iter().map(|&token| token as f32).collect();
    let candle_tensor = CandleTensor::from_vec(data, shape, &candle_core::Device::Cpu)?;

    Ok(ConcreteTensor::bitnet(candle_tensor))
}
```

#### Real Logits Extraction
```rust
/// Extract logits from neural network output tensor
fn tensor_to_logits(tensor: &ConcreteTensor, vocab_size: usize) -> Result<Vec<f32>> {
    match tensor {
        ConcreteTensor::BitNet(bitnet_tensor) => {
            let candle_tensor = bitnet_tensor.to_candle()?;

            // Handle different tensor shapes (batch_size, seq_len, vocab_size) or (vocab_size,)
            let logits = if candle_tensor.dims().len() == 3 {
                // Extract last position logits: [batch, seq, vocab] -> [vocab]
                let last_pos = candle_tensor.dims()[1] - 1;
                candle_tensor.get(0)?.get(last_pos)?
            } else if candle_tensor.dims().len() == 2 {
                // Extract last position: [seq, vocab] -> [vocab]
                let last_pos = candle_tensor.dims()[0] - 1;
                candle_tensor.get(last_pos)?
            } else {
                // Already in correct shape: [vocab]
                candle_tensor.clone()
            };

            // Convert to Vec<f32>
            let logits_vec = logits.to_vec1::<f32>()?;

            // Validate dimensions match expected vocab size
            if logits_vec.len() != vocab_size {
                return Err(anyhow::anyhow!(
                    "Logits dimension {} doesn't match vocab size {}. \
                     This indicates a model architecture mismatch.",
                    logits_vec.len(),
                    vocab_size
                ));
            }

            Ok(logits_vec)
        },
        ConcreteTensor::Mock(_) => {
            // Fallback for test environments
            Ok(vec![0.1; vocab_size])
        }
    }
}
```

### Alternative Approaches

1. **Direct Backend Integration**: Modify the backend interface to return logits directly
2. **Tensor Pipeline**: Create a dedicated tensor conversion pipeline
3. **Embeddings Integration**: Integrate with the embedding layer for token conversion

### Risk Assessment

**Implementation Risks:**
- Tensor shape mismatches between tokenizer output and model input
- Device compatibility issues (CPU/GPU tensor placement)
- Memory allocation failures with large vocabularies
- Integration issues with existing KV cache system

**Mitigation Strategies:**
- Comprehensive tensor shape validation
- Graceful fallback to mock tensors in test environments
- Device-aware tensor creation with proper error handling
- Integration tests with real models before deployment

### Phase 1: Token-to-Tensor Implementation (2-3 days)
- [ ] Replace `tokens_to_tensor` with real Candle tensor creation
- [ ] Add proper tensor shape validation
- [ ] Handle CPU/GPU device compatibility
- [ ] Add error handling for invalid token sequences

### Phase 2: Logits Extraction Implementation (2-3 days)
- [ ] Replace `tensor_to_logits` with real tensor data extraction
- [ ] Handle different output tensor shapes from various model architectures
- [ ] Add vocabulary size validation
- [ ] Optimize for performance with large vocabularies

### Phase 3: Integration Testing (1-2 days)
- [ ] Test with real GGUF models from the test suite
- [ ] Validate streaming generation quality improvement
- [ ] Performance benchmarking against mock implementation
- [ ] Cross-validation with non-streaming inference

### Phase 4: Error Handling & Edge Cases (1 day)
- [ ] Handle empty token sequences
- [ ] Manage tensor device mismatches
- [ ] Add comprehensive error messages
- [ ] Fallback strategies for unsupported configurations

### Unit Tests
```rust
#[test]
fn test_tokens_to_tensor_real_conversion() {
    let tokens = vec![101, 102, 103]; // Example token IDs
    let tensor = tokens_to_tensor(&tokens).unwrap();

    assert_eq!(tensor.shape(), &[1, 3]);
    match tensor {
        ConcreteTensor::BitNet(_) => {}, // Success
        ConcreteTensor::Mock(_) => panic!("Should create real tensor"),
    }
}

#[test]
fn test_tensor_to_logits_shape_validation() {
    let vocab_size = 50257;
    let mock_logits = vec![0.5; vocab_size];
    let tensor = create_test_tensor_from_logits(mock_logits);

    let extracted = tensor_to_logits(&tensor, vocab_size).unwrap();
    assert_eq!(extracted.len(), vocab_size);
    assert_ne!(extracted, vec![0.1; vocab_size]); // Not mock values
}
```

### Integration Tests
- Streaming generation with real models produces different outputs than mock
- Token probabilities vary meaningfully across vocabulary
- Generated text quality improves significantly
- Memory usage remains within acceptable bounds

### Performance Validation
- Latency impact of real tensor operations vs mocks
- Memory allocation patterns with different model sizes
- GPU utilization when using CUDA backend

### Core Functionality
- [ ] `tokens_to_tensor` creates valid `ConcreteTensor::BitNet` tensors from token arrays
- [ ] `tensor_to_logits` extracts real probability distributions from neural network outputs
- [ ] Streaming generation produces varied, contextually appropriate text instead of uniform output
- [ ] All existing streaming tests pass with real tensor operations

### Quality Improvements
- [ ] Generated text quality measurably improves over mock implementation
- [ ] Token probabilities reflect actual model predictions
- [ ] Different prompts produce different outputs (no longer deterministic)
- [ ] Stop sequences and EOS tokens function correctly with real logits

### Performance & Reliability
- [ ] Real tensor operations complete within streaming latency requirements (< 100ms per token)
- [ ] Memory usage remains stable during extended generation sessions
- [ ] Error handling gracefully manages tensor shape mismatches and device issues
- [ ] Fallback to mock tensors works correctly in test environments

### Integration
- [ ] Compatible with all existing backend implementations (CPU, GPU)
- [ ] Works with various model architectures and vocabulary sizes
- [ ] Integrates properly with KV cache and attention mechanisms
- [ ] Cross-validation with non-streaming inference produces consistent results

## Related Issues

This issue is part of a broader effort to eliminate mock implementations across the codebase:

- **Mock Objects Consolidation**: Extensive mock objects in streaming test modules need refinement
- **Backend Implementations**: Several backend methods contain stub implementations
- **Quantization Functions**: Multiple quantizer methods use simulation instead of real computation
- **Validation Framework**: Accuracy validation contains placeholder implementations

**Dependencies:**
- Requires stable `ConcreteTensor` and `BitNetTensor` APIs (already available)
- May benefit from enhanced error handling in tensor conversion utilities
- Should coordinate with ongoing backend implementation improvements

**Breaking Changes**: None expected - this is an internal implementation improvement that maintains the same public API.

---

**Priority**: Critical - blocks production streaming inference
**Effort**: Medium (4-6 days) - straightforward tensor operations with proper error handling
**Risk**: Low - well-defined tensor APIs with comprehensive fallback strategies
